flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/LogicalUnnestRule.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexNode;
+import org.immutables.value.Value;
+
+import java.util.Collections;
+import java.util.Map;
+
+import static org.apache.flink.table.types.logical.utils.LogicalTypeUtils.toRowType;
+
+/**
+ * Planner rule that rewrites UNNEST to explode function.
+ *
+ * <p>Note: This class can only be used in HepPlanner.
+ */
+@Value.Enclosing
+public class LogicalUnnestRule extends RelRule<LogicalUnnestRule.LogicalUnnestRuleConfig> {
+
+    public static final LogicalUnnestRule INSTANCE = LogicalUnnestRuleConfig.DEFAULT.toRule();
+
+    public LogicalUnnestRule(LogicalUnnestRule.LogicalUnnestRuleConfig config) {
+        super(config);
+    }
+
+    public boolean matches(RelOptRuleCall call) {
+        LogicalCorrelate join = call.rel(0);
+        RelNode right = getRel(join.getRight());
+        if (right instanceof LogicalFilter) {
+            LogicalFilter logicalFilter = (LogicalFilter) right;
+            RelNode relNode = getRel(logicalFilter.getInput());
+            if (relNode instanceof Uncollect) {
+                return !((Uncollect) relNode).withOrdinality;
+            } else if (relNode instanceof LogicalProject) {
+                LogicalProject logicalProject = (LogicalProject) relNode;
+                relNode = getRel(logicalProject.getInput());
+                if (relNode instanceof Uncollect) {
+                    return !((Uncollect) relNode).withOrdinality;
+                }
+                return false;
+            }
+        } else if (right instanceof LogicalProject) {
+            LogicalProject logicalProject = (LogicalProject) right;
+            RelNode relNode = getRel(logicalProject.getInput());
+            if (relNode instanceof Uncollect) {
+                Uncollect uncollect = (Uncollect) relNode;
+                return !uncollect.withOrdinality;
+            }
+            return false;
+        } else if (right instanceof Uncollect) {
+            Uncollect uncollect = (Uncollect) right;
+            return !uncollect.withOrdinality;
+        }
+        return false;
+    }
+
+    public void onMatch(RelOptRuleCall call) {
+        LogicalCorrelate correlate = call.rel(0);
+        RelNode outer = getRel(correlate.getLeft());
+        RelNode array = getRel(correlate.getRight());
+
+        // convert unnest into table function scan
+        RelNode tableFunctionScan = convert(array, correlate);
+        // create correlate with table function scan as input
+        Correlate newCorrelate =
+                correlate.copy(correlate.getTraitSet(), ImmutableList.of(outer, tableFunctionScan));
+        call.transformTo(newCorrelate);
+    }
+
+    private RelNode convert(RelNode relNode, LogicalCorrelate correlate) {
+        if (relNode instanceof HepRelVertex) {
+            HepRelVertex hepRelVertex = (HepRelVertex) relNode;
+            relNode = convert(getRel(hepRelVertex), correlate);
```

Comment:
it could happen that in some complex cases it will return `LogicalProject` or `LogicalFilter` or `Uncollect` and after that one of the further `ifs` will fire
That's why we can not just return here